src/runtime/timer.go

```diff
@@ -0,0 +1,157 @@
+package runtime
+
+import (
+	"internal/task"
+	"runtime/volatile"
+)
+
+// timer is an interface which can be implemented as a wrapper around a timer interrupt.
+type timer interface {
+	// setTimer configures the timer interrupt to fire when the ticks() value reaches the specified wakeup time.
+	// The timer interrupt must be disabled before calling this function.
+	setTimer(wakeup timeUnit)
+
+	// disableTimer disables the timer interrupt.
+	disableTimer()
+}
+
+// timerController is a type which manages a hardware timer, with builtin queueing.
+type timerController struct {
+	// fired is used to indicate whether the timer interrupt handler fired:
+	//   0: the timer interrupt handler has not fired
+	//   1: the timer interrupt handler fired
+	fired volatile.Register8
+
+	t timer
+
+	baseTime timeUnit
+	queue    *task.Task
+}
+
+// set configures the timer interrupt to fire when the next timer expires.
+// This function assumes that there is a timer on the queue.
+func (c *timerController) set() {
+	// Disable the timer to avoid a race condition.
+	c.t.disableTimer()
+
+	// Clear the fired flag.
+	c.fired.Set(0)
+
+	// Set the timer with the appropriate time.
+	c.t.setTimer(c.baseTime)
+}
+
+// enqueue queues a task to be awoken at the specified wakeup time.
+func (c *timerController) enqueue(t *task.Task, wakeup timeUnit) {
+	t.Data = 0
+
+	// Insert task into timer queue.
+	var baseUpdated bool
+	switch {
+	case c.queue == nil:
+		// There were no timers previously in the queue.
+		scheduleLog("  new timer queue")
+		c.queue = t
+		c.baseTime = wakeup
+		baseUpdated = true
+	case wakeup < c.baseTime:
+		// This task comes before everything else on the queue.
+		c.queue.Data += uint(c.baseTime - wakeup)
+		t.Next = c.queue
+		c.queue = t
+		c.baseTime = wakeup
+		baseUpdated = true
+	default:
+		// Insert this task later in the queue.
+		offset := wakeup - c.baseTime
+		prev := c.queue
+		for ; prev.Next != nil && offset < timeUnit(prev.Next.Data); prev = prev.Next {
+			offset -= timeUnit(prev.Data)
+		}
+		prev.Next, t.Next = t, prev.Next
+		if t.Next != nil {
+			t.Next.Data -= uint(offset)
+		}
+		t.Data = uint(offset)
+	}
+
+	if c.fired.Get() != 0 {
+		// The timer already fired.
+		// Poll for completed timer events.
+		c.poll()
+		return
+	}
+
+	if baseUpdated {
+		// The base time has changed, so the timer interrupt needs to be reconfigured.
+		c.set()
+	}
+}
+
+// poll checks for any expired timers and schedules them.
+// This must be called periodically by the scheduler.
```

Comment:
This feels like a possible problem. Why is there a need for polling?

Reply:
We have 2 kinds of platforms in TinyGo:
1. microcontrollers -  where all events come from interrupts
2. "hosted" platforms (wasm, embedded linux, etc.) - where we need to periodically poll for events

We *could* use 2 different scheduler implementations, one interrupt based and one polling, but that would make some stuff more complicated and result in 2 different sets of behavior.

Reply:
What kinds of events need to be polled for on hosted platforms? Epoll?
I may need to look more into it to understand the problem better.